mypy/checker.py

```diff
@@ -4118,6 +4118,61 @@ def has_no_custom_eq_checks(t: Type) -> bool:
                             expr_indices,
                             narrowable_operand_index_to_hash.keys(),
                         )
+
+                    def is_type_call(expr: CallExpr) -> bool:
+                        """Is expr a call to type with one argument?"""
+                        return (refers_to_fullname(expr.callee, 'builtins.type')
+                                and len(expr.args) == 1)
+
+                    # If we haven't been able to narrow types yet, we might be dealing with a
+                    # explicit type(x) == some_type check
+                    if if_map == {} and else_map == {}:
+                        # exprs that are being passed into type
+                        exprs_in_type_calls = []  # type: List[Expression]
+                        # type that is being compared to type(expr)
+                        type_being_compared = None  # type: Optional[List[TypeRange]]
+
+                        for index in expr_indices:
+                            expr = node.operands[index]
+
+                            if isinstance(expr, CallExpr) and is_type_call(expr):
+                                exprs_in_type_calls.append(expr.args[0])
+                            else:
+                                current_type = get_isinstance_type(expr, type_map)
+                                if current_type is None:
+                                    continue
+                                if type_being_compared is not None:
+                                    # It doesn't really make sense to have several types being
+                                    # compared to the output of type (like type(x) == int == str)
+                                    # because whether that's true is solely dependent on what the
+                                    # types being compared are, so we don't try to narrow types any
+                                    # further because we can't really get any information about the
+                                    # type of x from that check
+                                    break
+                                else:
+                                    type_being_compared = current_type
+
+                        if_maps = []  # type: List[TypeMap]
```

Comment:
Can you move the remaining operations behind `if exprs_in_type_calls:`, as an optimization? This way the remaining code will only be run if there is something to do. It also makes it easier to see that the code does nothing unless there is a `type(x)` somewhere.